### DISCUSSION TITLE
Question about PlatformIO and how to upload Rules to ESP8266EX

### DISCUSSION THREAD

**Discussion by jackpotcityco** [arendst/Tasmota#19721] [Q&A] [answered]
Hello,

I am new to Board development. I will state what I have done below before the question.

1. I have a ESP8266EX board with Tasmota Firmware. I also have a "vl53l0x" distance sensor.
2. In the web interface on "http://15.1.0.14/", I can see the distance in the Console updating the values.
3. I have added this Rule in the console for that: 
    **Rule1 ON VL53L0X#Distance DO ConsoleLog,Distance:%value% ENDON**
4. I have installed "Visual Studio Code" with "PlatformIO". I also have installed under Platforms: "**Espressif 8266**"
5. I have the ESP8266EX board connected with a microUSB as I can see the values in the web interface.
6. I have installed the driver: "USB-SERIAL CH340 (COM3) which is seen in Device Manager.

My goal is to upload "Rule4" using the below Python script but the code gets stuck on below line. I am not sure what could be wrong.
**result = subprocess.run(upload_command, shell=False, stdout=subprocess.PIPE, stderr=subprocess.PIPE)**

`import os
import subprocess

print("Script is running1")

rule = "Rule4 ON VL53L0X#Distance DO ConsoleLog,Distance:%value% ENDON"
filename_to_upload = "data/rule.txt"

print("Script is running1b")

with open(filename_to_upload, "w") as file:
    file.write(rule)
upload_command = ["pio", "run", "--target", "uploadfs"]
result = subprocess.run(upload_command, shell=False, stdout=subprocess.PIPE, stderr=subprocess.PIPE)

if result.returncode == 0:
    print(f"{filename_to_upload} uploaded successfully.")
else:
    print(f"Error uploading {filename_to_upload}:", result.stderr.decode())`

**The platformio.ini file is the below:**
`[env:upload_rule]
platform = espressif8266
board = esp07
framework = arduino
upload_port = COM3  ; Replace COM3 with your ESP8266's actual COM port

; Custom targets for uploading rule and file system
extra_scripts = pre:upload_rule.py
targets = uploadfs`

Thank you!

**Comment by barbudor** [comment]
I have no idea what you are trying to do with platformio, but there is no 'Rule4' on Tasmota
It stops at 3 for now

**Reply by jackpotcityco** [reply]
I am very new to this board, so this is my first code with boards and Rules. I come from C# development normally.

I see, so the ESP8266 can only have 3 rules. Is that true? Does it mean that it is the maximum of rules even if having more than a distance sensor connected to the ESP8266EX ?

**Reply by barbudor** [reply]
So you seem to try uploading a file to the filesystem, but you don't say which Tasmota variant on which board.
Standard tasmota build is meant for ESP with 1MB with no filesystem configured.

**Reply by barbudor** [reply]
I wonder why for your first contact with tasmota you are already diving into the internals of platformio...

Nevertheless, read the doc about rules.
There are 3 rule-sets (Rule1 to rule3) which each can contain as many rules that the 512 bytes buffer can hold considering that it can be compressed. So a capacity of around 1000 to 1400 characters by rule-set

**Comment by jackpotcityco** [comment]
**>but you don't say which Tasmota variant on which board.**
I was handed this board from someone so I am not exactly sure. The only thing I can see on the board is on the chip:
**ESP8266EX
042021
JB00POF109**
The Program Version 13.1.0.4(tasmota) is what I can see in the web interface. 

That might not be enough information perheps.

**>I wonder why for your first contact with tasmota you are already diving into the internals of platformio**
Only reason would be to be able to upload/change rules with python scripts as I later would need to dynamically change the rules

**>There are 3 rule-sets (Rule1 to rule3) which each can contain as many rules that the 512 bytes buffer can hold considering that it can be compressed. So a capacity of around 1000 to 1400 characters by rule-set**
That was interesting. I must go and look for how that would be possible to do. Perheps I could find a way here also how to do that.

I am sorry for beginner questions.

**Reply by sfromis** [reply, marked as answer]
If you want to programmatically upload rule sets, you have to do it by issuing Tasmota console commands `Rule1` / `Rule2` / `Rule3`. That can be done by various means, like MQTT, web console, http commands or serial console. It is not as files.

If you want to do a build with rule sets baked in, you can also set `#define` for `USER_RULE1` etc. And also `USER_BACKLOG` if you want to integrate commands to also activate the rule sets, executed once first boot when initializing from scratch.

And see https://tasmota.github.io/docs/Rules/

**Comment by jackpotcityco** [comment]
>by issuing Tasmota console commands Rule1 / Rule2 / Rule3

Thank you, I might start to perheps use those to get a basic understanding first and try with the other things you mentioned later then.

Thank you for your help!
